toolchain/check/check_unit.cpp

```diff
@@ -570,24 +570,112 @@ auto CheckUnit::CheckPoisonedConcreteImplLookupQueries() -> void {
   context_.inst_block_stack().PopAndDiscard();
 }
 
+// Check for invalid overlap between impls, given the set of all impls for a
+// single interface.
+static auto CheckOverlappingImplsForInterface(
+    Context& context,
+    llvm::ArrayRef<std::pair<SemIR::ImplId, const SemIR::Impl*>> impls)
+    -> void {
+  // TODO: We should revisit this and look for a way to do these checks in less
+  // than quadratic time. From @zygoloid: Possibly by converting the set of
+  // impls into a decision tree.
+  for (auto [index_a, impl_a_pair] : llvm::enumerate(impls)) {
+    auto [impl_a_id, impl_a] = impl_a_pair;
+    if (impl_a->witness_id == SemIR::ErrorInst::InstId) {
+      continue;
+    }
+    auto type_structure =
+        BuildTypeStructure(context, impl_a->self_id, impl_a->interface);
+
+    for (const auto& impl_b_pair : llvm::drop_begin(impls, index_a + 1)) {
+      auto [impl_b_id, impl_b] = impl_b_pair;
+      if (impl_b->witness_id == SemIR::ErrorInst::InstId) {
+        continue;
+      }
+
+      // The type structure each non-final `impl` must differ from all other
+      // non-final `impl` for the same interface visible from the file.
+      if (!impl_a->is_final && !impl_b->is_final) {
+        auto type_structure2 =
+            BuildTypeStructure(context, impl_b->self_id, impl_b->interface);
+        if (type_structure == type_structure2) {
+          CARBON_DIAGNOSTIC(ImplNonFinalSameTypeStructure, Error,
+                            "found non-final `impl` with the same type "
+                            "structure as another non-final `impl`");
+          auto builder = context.emitter().Build(impl_b->latest_decl_id(),
+                                                 ImplNonFinalSameTypeStructure);
+          CARBON_DIAGNOSTIC(ImplNonFinalSameTypeStructureNote, Note,
+                            "other `impl` here");
+          builder.Note(impl_a->latest_decl_id(),
+                       ImplNonFinalSameTypeStructureNote);
+          builder.Emit();
+          break;
+        }
+      } else {
+        CARBON_CHECK(impl_a->is_final || impl_b->is_final);
+
+        auto diagnose = [&](const SemIR::Impl& query_impl,
+                            SemIR::ImplId final_impl_id,
+                            const SemIR::Impl& final_impl) -> bool {
+          if (LookupMatchesImpl(
```

Comment:
Is it possible that this would perform any importing? (I'm thinking more about this due to #5438.) I think `LookupMatchesImpl` can perform deduction, which can perform conversion, which can import, so this call can invalidate all the `Impl*`s and `Impl&`s used in this function. We might need to switch to only dealing in `ImplId`s here.

Reply:
Oh no.. yeah, thank you. That ASAN poisoning is going to pay dividends until we get self-host :)

Reply:
Done